### PR TITLE
Develop

### PR DESCRIPTION
**Pull request by ruancomelli**
# Description

_Explain briefly about the purpose & context of this PR_

# Screenshots

_If possible add a screenshot and/or gif_

# Example | Usage

_Please share examples and how to reproduce_

---

# Related Issues | Pr

_`close`|`resolve` <ISSUE> (ex. `resolve #20`)_
_link issues to close automatically_

### CODE REVIEW

**Comment by ruancomelli** [comment]
This is an experiment review for experiment `multiple_mermaid_diagram_types`
Run ID: `multiple_mermaid_diagram_types/run_2024-09-26T16-03-48_v1-23-0-21-g3eb85d2d2`
The benchmark review for this pull request can be found at https://github.com/sourcery-ai-experiments/timba-api/pull/12.
This pull request was cloned from `https://github.com/AlphaDev87/timba-api/pull/49`. _(Note: the URL is not a link to avoid triggering a notification on the original pull request.)_

<details>
<summary>Experiment configuration</summary>

```yaml
review_config:
  # User configuration for the review
  # - benchmark - use the user config from the benchmark reviews
  # - <value> - use the value directly
  user_review_config:
    enable_ai_review: true
    enable_rule_comments: false

    enable_complexity_comments: false
    enable_security_comments: false
    enable_tests_comments: false
    enable_comment_suggestions: false
    enable_functionality_review: false

    enable_pull_request_summary: false
    enable_review_guide: true

    enable_approvals: false

  ai_review_config:
    # The model responses to use for the experiment
    # - benchmark - use the model responses from the benchmark reviews
    # - llm - call the language model to generate responses
    model_responses:
      comments_model: benchmark
      comment_area_model: benchmark
      comment_validation_model: benchmark
      comment_suggestion_model: benchmark
      complexity_model: benchmark
      functionality_model: benchmark
      security_model: benchmark
      tests_model: benchmark
      pull_request_summary_model: benchmark
      review_guide_model: llm
      overall_comments_model: benchmark

# The pull request dataset to run the experiment on
pull_request_dataset:
- https://github.com/sourcery-ai/core/pull/4607
- https://github.com/sourcery-ai/core/pull/4631
- https://github.com/sourcery-ai/core/pull/4647
    # CodeRabbit examples:
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/alexsnow348/insightface/pull/46
- https://github.com/alikuxac/utilities/pull/10
- https://github.com/AlphaDev87/timba-api/pull/49
- https://github.com/AngeloTadeucci/Maple2/pull/239
- https://github.com/AngeloTadeucci/Maple2.File/pull/36
- https://github.com/AngeloTadeucci/Maple2/pull/233
    # Examples where CodeRabbit does not generate diagrams
- https://github.com/baptisteArno/typebot.io/pull/1778
- https://github.com/btipling/foundations/pull/33
- https://github.com/btipling/foundations/pull/31
- https://github.com/chintu-777/jaeger/pull/1
- https://github.com/coji/remix-docs-ja/pull/55
- https://github.com/DaveMBush/SmartNgRX/pull/622
- https://github.com/DaveMBush/SmartNgRX/pull/481
- https://github.com/dkittle/party-connections/pull/6
- https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
- https://github.com/imaami/libcanth/pull/2
    # Requested by Tim
- https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
- https://github.com/2lambda123/galaxyproject-galaxy/pull/12
- https://github.com/a0v0/avtoolz/pull/79
- https://github.com/adityask98/Hotaru/pull/10
- https://github.com/agdas/vscode/pull/2
- https://github.com/agluszak/hirschgarten/pull/2
- https://github.com/4DNucleome/PartSeg/pull/1183
- https://github.com/ably/ably-cocoa/pull/1973
- https://github.com/ably-labs/ably-chat-swift/pull/54
- https://github.com/ably-labs/ably-chat-swift/pull/35
  # Including database changes to evaluate entity-relationship diagrams
- https://github.com/sourcery-ai/core/pull/4579
  # TODO: find some with Alembic

# Questions to ask to label the review comments
review_comment_labels: []
# - label: correct
#   question: Is this comment correct?

# Benchmark reviews generated by running
#   python -m scripts.experiment benchmark <experiment_name>
benchmark_reviews:
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4607
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/374
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4631
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/375
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4647
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/376
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/5
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/13
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/14
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/14
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/13
- dataset_pull_request: https://github.com/alikuxac/utilities/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/utilities/pull/4
- dataset_pull_request: https://github.com/AlphaDev87/timba-api/pull/49
  review_pull_request: https://github.com/sourcery-ai-experiments/timba-api/pull/12
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/239
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/23
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2.File/pull/36
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2.File/pull/12
- dataset_pull_request: https://github.com/AngeloTadeucci/Maple2/pull/233
  review_pull_request: https://github.com/sourcery-ai-experiments/Maple2/pull/24
- dataset_pull_request: https://github.com/baptisteArno/typebot.io/pull/1778
  review_pull_request: https://github.com/sourcery-ai-experiments/typebot.io/pull/7
- dataset_pull_request: https://github.com/btipling/foundations/pull/33
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/13
- dataset_pull_request: https://github.com/btipling/foundations/pull/31
  review_pull_request: https://github.com/sourcery-ai-experiments/foundations/pull/14
- dataset_pull_request: https://github.com/chintu-777/jaeger/pull/1
  review_pull_request: https://github.com/sourcery-ai-experiments/jaeger/pull/7
- dataset_pull_request: https://github.com/coji/remix-docs-ja/pull/55
  review_pull_request: https://github.com/sourcery-ai-experiments/remix-docs-ja/pull/7
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/622
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/13
- dataset_pull_request: https://github.com/DaveMBush/SmartNgRX/pull/481
  review_pull_request: https://github.com/sourcery-ai-experiments/SmartNgRX/pull/14
- dataset_pull_request: https://github.com/dkittle/party-connections/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/party-connections/pull/7
- dataset_pull_request: https://github.com/Drajad-Kusuma-Adi/onstudy-backend/pull/6
  review_pull_request: https://github.com/sourcery-ai-experiments/onstudy-backend/pull/7
- dataset_pull_request: https://github.com/imaami/libcanth/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/libcanth/pull/7
- dataset_pull_request: https://github.com/2lambda123/-Orange-OpenSource-oorobot/pull/15
  review_pull_request: https://github.com/sourcery-ai-experiments/-Orange-OpenSource-oorobot/pull/6
- dataset_pull_request: https://github.com/2lambda123/galaxyproject-galaxy/pull/12
  review_pull_request: https://github.com/sourcery-ai-experiments/galaxyproject-galaxy/pull/14
- dataset_pull_request: https://github.com/adityask98/Hotaru/pull/10
  review_pull_request: https://github.com/sourcery-ai-experiments/Hotaru/pull/15
- dataset_pull_request: https://github.com/agdas/vscode/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/vscode/pull/15
- dataset_pull_request: https://github.com/agluszak/hirschgarten/pull/2
  review_pull_request: https://github.com/sourcery-ai-experiments/hirschgarten/pull/14
- dataset_pull_request: https://github.com/4DNucleome/PartSeg/pull/1183
  review_pull_request: https://github.com/sourcery-ai-experiments/PartSeg/pull/7
- dataset_pull_request: https://github.com/ably/ably-cocoa/pull/1973
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-cocoa/pull/6
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/54
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/11
- dataset_pull_request: https://github.com/ably-labs/ably-chat-swift/pull/35
  review_pull_request: https://github.com/sourcery-ai-experiments/ably-chat-swift/pull/12
- dataset_pull_request: https://github.com/sourcery-ai/core/pull/4579
  review_pull_request: https://github.com/sourcery-ai-experiments/core/pull/377

```

</details>

**Comment by SourceryAI** [comment]
## Reviewer's Guide by Sourcery

This pull request implements significant changes to the project structure, database schema, and functionality. It includes new features like password reset, analytics, and improvements to existing components such as deposits, payments, and user management. The changes also involve refactoring of existing code, addition of new API endpoints, and updates to the error handling and validation mechanisms.

### Sequence Diagrams

#### Sequence diagram for password reset process

```mermaid
sequenceDiagram
    participant U as User
    participant A as Auth Router
    participant N as NewPasswordController
    participant S as NewPasswordServices
    participant E as Email Service
    participant D as Database

    U->>A: POST /auth/forgot-password
    A->>N: create()
    N->>S: forgotPassword(username)
    S->>D: Get user by username
    S->>S: Generate reset token
    S->>D: Save reset token
    S->>E: Send reset email
    N-->>U: Success response

    U->>A: POST /auth/restore-password
    A->>N: store()
    N->>S: resetPassword(user, new_password)
    S->>D: Update user password
    S->>D: Delete reset token
    N-->>U: Success response
```

#### Class diagram for error handling

```mermaid
classDiagram
    class CustomError {
        +status: number
        +code: string
        +description: string | ValidationError[]
        +detail?: object
        +constructor(err: ErrorData)
    }
    class AgentApiError {
        +constructor(status: number, description: string, detail: any)
    }
    class AlquimiaApiError {
        +constructor(status: number, description: string, detail: any)
    }
    CustomError <|-- AgentApiError
    CustomError <|-- AlquimiaApiError
```

### File-Level Changes

| Change | Details | Files |
| ------ | ------- | ----- |
| Implement password reset functionality | <ul><li>Add new password reset endpoints</li><li>Create NewPasswordServices for handling password reset logic</li><li>Implement email sending for password reset</li><li>Add rate limiting for password reset requests</li></ul> | `src/routes/auth.router.ts`<br/>`src/components/new-password/controller.ts`<br/>`src/components/new-password/services.ts`<br/>`src/components/new-password/validators.ts`<br/>`src/db/password-reset-token.ts`<br/>`src/helpers/email/email.ts`<br/>`src/middlewares/rate-limiters/pass-restore-request.ts` |
| Refactor and improve deposits and payments handling | <ul><li>Split transactions router into separate deposits and payments routers</li><li>Create new DepositServices and PaymentServices</li><li>Implement new endpoints for listing and managing deposits and payments</li><li>Add pagination and sorting capabilities to deposit and payment listings</li></ul> | `src/routes/transactions/deposits.router.ts`<br/>`src/routes/transactions/payments.router.ts`<br/>`src/components/deposits/controller.ts`<br/>`src/components/deposits/services.ts`<br/>`src/components/payments/controller.ts`<br/>`src/components/payments/services.ts`<br/>`src/db/deposits.ts`<br/>`src/db/payments.ts` |
| Add analytics functionality | <ul><li>Create new analytics router and controller</li><li>Implement endpoints for creating and retrieving analytics data</li><li>Add new ANALYTICS table to the database schema</li></ul> | `src/routes/analytics.router.ts`<br/>`src/components/analytics/controller.ts`<br/>`src/components/analytics/validators.ts`<br/>`src/db/analytics.ts`<br/>`prisma/migrations/20240605135628_create_analytics_table/migration.sql`<br/>`prisma/migrations/20240605142957_rename_analytics_owner_to_source/migration.sql` |
| Improve error handling and validation | <ul><li>Refactor CustomError class and error handling middleware</li><li>Implement new error types like AgentApiError and AlquimiaApiError</li><li>Update validation schemas for various endpoints</li></ul> | `src/helpers/error/CustomError.ts`<br/>`src/helpers/error/AgentApiError.ts`<br/>`src/helpers/error/AlquimiaApiError.ts`<br/>`src/middlewares/errorHandler.ts`<br/>`src/components/players/validators.ts`<br/>`src/components/agent/validators.ts` |
| Implement new bot-related functionality | <ul><li>Add new bot router and controller</li><li>Implement endpoint for retrieving bot QR codes</li><li>Create CHAT_DATA table in the database</li></ul> | `src/routes/bot.router.ts`<br/>`src/components/bot/controller.ts`<br/>`src/components/bot/services.ts`<br/>`src/components/bot/validators.ts`<br/>`prisma/migrations/20240514142314_create_chat_data_table/migration.sql`<br/>`prisma/migrations/20240514143050_add_unique_constraint_to_chat_data_number/migration.sql` |
| Update database schema and migrations | <ul><li>Implement new database tables and relationships</li><li>Update existing tables with new columns and constraints</li><li>Remove deprecated tables and columns</li></ul> | `prisma/migrations/20240508181836_initial_migration/migration.sql` |
| Refactor and improve existing components | <ul><li>Update player management functionality</li><li>Improve agent-related operations</li><li>Refactor authentication and authorization mechanisms</li></ul> | `src/components/players/controller.ts`<br/>`src/components/players/services.ts`<br/>`src/components/agent/controller.ts`<br/>`src/components/agent/services.ts`<br/>`src/components/auth/services.ts` |

---

<details>
<summary>Tips and commands</summary>

#### Interacting with Sourcery

- **Trigger a new review:** Comment `@sourcery-ai review` on the pull request.
- **Continue discussions:** Reply directly to Sourcery's review comments.
- **Generate a GitHub issue from a review comment:** Ask Sourcery to create an
  issue from a review comment by replying to it.

#### Customizing Your Experience

Access your [dashboard](https://app.sourcery.ai) to:
- Enable or disable review features such as the Sourcery-generated pull request
  summary, the reviewer's guide, and others.
- Change the review language.
- Add, remove or edit custom review instructions.
- Adjust other review settings.

#### Getting Help

- [Contact our support team](mailto:support@sourcery.ai) for questions or feedback.
- Visit our [documentation](https://docs.sourcery.ai) for detailed guides and information.
- Keep in touch with the Sourcery team by following us on [X/Twitter](https://x.com/SourceryAI), [LinkedIn](https://www.linkedin.com/company/sourceryai/) or [GitHub](https://github.com/sourcery-ai).

</details>

File: src/routes/agent.router.ts [hunk 1]
```diff
@@ -67,5 +54,19 @@ agentRouter.post(
...
+  checkExact(),
+  throwIfBadRequest,
+  AgentController.updateSupportNumbers,
+);
+agentRouter.post(
+  "/reset-player-password",
+  validateResetPasswordRequest(),
+  checkExact(),
+  throwIfBadRequest,
+  AgentController.resetPlayerPassword,
```
Comment (SourceryAI) [inline src/routes/agent.router.ts:70]: **🚨 suggestion (security):** New password reset functionality needs security review

The addition of a password reset functionality for agents is a sensitive operation. Ensure that this endpoint is properly secured, audited, and only accessible to authorized agents. Consider implementing additional safeguards such as requiring secondary authentication for this operation.

File: src/db/payments.ts [hunk 2]
```diff
@@ -111,4 +141,30 @@ export class PaymentsDAO {
       prisma.$disconnect();
     }
   }
+
+  static async authorizeRelease(
```
Comment (SourceryAI) [inline src/db/payments.ts:145]: **suggestion:** Critical operation for payment release, needs careful review

The authorizeRelease method is a critical operation for payment processing. Ensure that all edge cases are handled, such as concurrent requests or system failures during the transaction. Consider adding more detailed logging or an audit trail for this operation.

File: src/db/players.ts [hunk 3]
```diff
@@ -1,26 +1,73 @@
...
+    try {
+      return prisma.player.count();
+    } catch (e) {
+      throw e;
+    } finally {
+      prisma.$disconnect();
+    }
+  }
+
+  static _getAll = async (
```
Comment (SourceryAI) [inline src/db/players.ts:30]: **suggestion (performance):** New pagination and search methods need performance optimization

The addition of pagination and search functionality is good for scalability. However, ensure that these queries are optimized for performance, especially for large datasets. Consider adding indexes on frequently searched fields and test the performance with a large number of records.

**Review by SourceryAI — COMMENTED** [review]
Hey @ruancomelli - I've reviewed your changes and they look great!

<details>
<summary>Here's what I looked at during the review</summary>

- 🟡 **General issues**: 2 issues found
- 🟢 **Documentation**: all looks good
</details>

***

[LangSmith trace](https://smith.langchain.com/o/7d8518ad-6dd3-5c7b-96d2-757769296335/projects/p/88560225-5072-48bc-967c-dc54da53f265/r/ac9434d0-2b6b-401f-8396-a076b0824baf?poll=true)

<sub>
Help me be more useful! Please click 👍 or 👎 on each comment to tell me if it was helpful.
</sub>

**Pull request closed without merging**